File: src/anyware/window.clj
(ns anyware.window
  (:require [clojure.string :as string]
            [anyware.buffer :as buffer]))

(defrecord Window [key x y width height])

(defn html [buffers window]
  (let [buffer (get buffers (:key window))
        lines (string/split (buffer/text buffer) \newline)
        length (count lines)]
    (subvec (vec (map (fn [line] (let [length (count line)] (subs line (min (:x window) length) (min (:width window) length)))) lines)) (min (:y window) length) (min (:height window) length))))

File: src/main/scala/anyware/Window.scala
package anyware

import scala.xml.Node

case class Window(name: String, x: Int, y: Int, width: Int, height: Int) {

  def toHTML(buffers: Map[String, Buffer]): Node =
    <pre><code>{buffers(name).toHTML.slice(y, y + height).padTo(height, <br />)}</code></pre>

}

object Window {

  def create(name: String): Window = Window(name, 0, 0, 0, 0)

}
